cdc/kv/region_worker.go

```diff
@@ -400,38 +412,42 @@ func (w *regionWorker) processEvent(ctx context.Context, event *regionStatefulEv
 		return nil
 	}
 	var err error
-	event.state.lock.Lock()
 	if event.changeEvent != nil {
 		w.metrics.metricReceivedEventSize.Observe(float64(event.changeEvent.Event.Size()))
 		switch x := event.changeEvent.Event.(type) {
 		case *cdcpb.Event_Entries_:
+			event.state.lock.Lock()
 			err = w.handleEventEntry(ctx, x, event.state)
 			if err != nil {
 				err = w.handleSingleRegionError(err, event.state)
 			}
+			event.state.lock.Unlock()
 		case *cdcpb.Event_Admin_:
 			log.Info("receive admin event",
 				zap.Stringer("event", event.changeEvent),
 				zap.String("namespace", w.session.client.changefeed.Namespace),
 				zap.String("changefeed", w.session.client.changefeed.ID))
 		case *cdcpb.Event_Error:
+			event.state.lock.Lock()
 			err = w.handleSingleRegionError(
 				cerror.WrapError(cerror.ErrEventFeedEventError, &eventError{err: x.Error}),
 				event.state,
 			)
+			event.state.lock.Unlock()
 		case *cdcpb.Event_ResolvedTs:
-			if err = w.handleResolvedTs(ctx, x.ResolvedTs, event.state); err != nil {
-				err = w.handleSingleRegionError(err, event.state)
-			}
+			err = w.handleResolvedTs(ctx, &resolvedTsEvent{
+				resolvedTs: x.ResolvedTs,
+				regions:    []*regionFeedState{event.state},
+			})
+		}
+		if err != nil {
+			return err
 		}
 	}
 
-	if event.resolvedTs != nil {
-		if err = w.handleResolvedTs(ctx, event.resolvedTs.Ts, event.state); err != nil {
-			err = w.handleSingleRegionError(err, event.state)
```

Comment:
why need handleSingleRegionError function in the older version and remove it now ?

Reply:
handleResolvedTs throw error only if the context is done